### PR TITLE
Optimize CBezierCurve::getYForPoint

### PR DESCRIPTION
**Pull request by Dickby**
#### Describe your PR, what does it fix/add?
As the title sais, it optimizes getYForPoint

#### Is there anything you want to mention? (unchecked code, possible bugs, found problems, breaking compatibility, etc.)
I tested this against the former solution, 
without some tricks the benchmark part of the bezier setup is optimized out.
There is no functional difference to the replaced code other than performance.

#### Is it ready for merging, or does it need work?
Someone should check the performance on other computers.

### CODE REVIEW

**Comment by Dickby** [comment]
What's up with this Nix shit?
I am not responsible, am I?

**Comment by Dickby** [comment]
Playing around with this, i reallized that the old implementation could lead to lowerT == upperT after clamping.
Of course, otherwise clamping would not be needed.
That could either happen for lowerT = upperT = 0, or lowerT = upperT = BAKEDPOINTS -1.
In Both cases PERCINDELTA would become nan i think.
After the latest change to my PR this should not happen anymore.

**Comment by Dickby** [comment]
A thing to think about:
For our use case, bezier curves that have more than one y value for a given x value don't make sense.
They could lead to jumps in animations.
It would be nice to restrict the possible bezier curves so that every getXForT(t) < getXForT(t+epsylon).

File: src/helpers/BezierCurve.cpp [hunk 1]
```diff
@@ -46,34 +46,34 @@ float CBezierCurve::getXForT(float t) {
...
-        mid = (upperT + lowerT) / 2;
+        below = m_aPointsBaked[index].x < x;
     }
 
-    // in the name of performance i shall make a hack
-    const auto LOWERPOINT = &m_aPointsBaked[std::clamp(lowerT, 0, BAKEDPOINTS - 1)];
-    const auto UPPERPOINT = &m_aPointsBaked[std::clamp(upperT, 0, BAKEDPOINTS - 1)];
+    int lowerIndex = index - (!below || index == BAKEDPOINTS - 1);
+
+    // in the name of performance i shall make a hack (is the hack still here or can this comment be deleted?)
```
Comment (vaxerski) [inline src/helpers/BezierCurve.cpp:68]: it is, don't remove the comment :)

File: src/helpers/BezierCurve.cpp [hunk 2]
```diff
@@ -46,34 +46,34 @@ float CBezierCurve::getXForT(float t) {
...
-
-    while (std::abs(upperT - lowerT) > 1) {
-        if (m_aPointsBaked[mid].x > x) {
-            upperT = mid;
+    if (x >= 1.f)
+        return 1.f;
+
+    int  index = 0;
+    bool below = true;
+    for (int step = (BAKEDPOINTS + 1) / 2; step; step /= 2) {
```
Comment (vaxerski) [inline src/helpers/BezierCurve.cpp:57]: I'd rather have some more sane condition for step like e.g. step > 0

File: src/helpers/BezierCurve.cpp [hunk 3]
```diff
@@ -46,34 +46,34 @@ float CBezierCurve::getXForT(float t) {
...
+        if (below) {
```
Comment (vaxerski) [inline src/helpers/BezierCurve.cpp:58]: I know old code had the issue too, but no {} around short ifs

**Comment by vaxerski** [comment]
> For our use case, bezier curves that have more than one y value for a given x value don't make sense.
They could lead to jumps in animations.
It would be nice to restrict the possible bezier curves so that every getXForT(t) < getXForT(t+epsylon).

Indeed.

**Comment by vaxerski** [comment]
I will review this tomorrow, as I need to do some analysis

**Review by vaxerski — APPROVED** [review]
lgtm :) thanks!

**Pull request merged**